File: sdk/android/story-sdk/src/main/kotlin/com/openstory/sdk/internal/cache/StoryFeedCacheKey.kt
package com.openstory.sdk.internal.cache

import com.openstory.sdk.internal.context.UserSegmentsNormalizer
import java.security.MessageDigest

internal data class StoryFeedCacheKey(
    val placementKey: String,
    val platform: String,
    val appVersion: String,
    val normalizedUserSegments: List<String>,
    val userSegmentsHash: String,
) {
    val databaseKey: String =
        "$placementKey::$platform::$appVersion::$userSegmentsHash"

    companion object {
        fun create(
            placementKey: String,
            platform: String,
            appVersion: String,
            userSegments: Collection<String>,
        ): StoryFeedCacheKey {
            val normalizedSegments = UserSegmentsNormalizer.normalize(userSegments)
            val digest = MessageDigest.getInstance("SHA-256")
            val bytes = digest.digest(normalizedSegments.joinToString("\n").toByteArray())
            val hash = bytes.joinToString(separator = "") { byte -> "%02x".format(byte) }

            return StoryFeedCacheKey(
                placementKey = placementKey.trim(),
                platform = platform.trim(),
                appVersion = appVersion.trim(),
                normalizedUserSegments = normalizedSegments,
                userSegmentsHash = hash,
            )
        }
    }
}

File: sdk/flutter/ios/open_story_flutter/Sources/open_story_flutter/OpenStorySDK/Internal/Cache/StoryFeedCacheKey.swift
import CryptoKit
import Foundation

internal struct StoryFeedCacheKey: Equatable {
    let placementKey: String
    let platform: String
    let appVersion: String
    let normalizedUserSegments: [String]
    let userSegmentsHash: String

    var databaseKey: String {
        "\(placementKey)::\(platform)::\(appVersion)::\(userSegmentsHash)"
    }

    static func create(
        placementKey: String,
        platform: String,
        appVersion: String,
        userSegments: some Collection<String>
    ) -> StoryFeedCacheKey {
        let normalizedUserSegments = UserSegmentsNormalizer.normalize(userSegments)
        let digest = SHA256.hash(
            data: Data(normalizedUserSegments.joined(separator: "\n").utf8)
        )
        let hash = digest.map { String(format: "%02x", $0) }.joined()

        return StoryFeedCacheKey(
            placementKey: placementKey.trimmingCharacters(in: .whitespacesAndNewlines),
            platform: platform.trimmingCharacters(in: .whitespacesAndNewlines),
            appVersion: appVersion.trimmingCharacters(in: .whitespacesAndNewlines),
            normalizedUserSegments: normalizedUserSegments,
            userSegmentsHash: hash
        )
    }
}
